### PR TITLE
template starter

### PR DESCRIPTION
**Pull request by zeeshanlakhani**
Implements most of https://github.com/fission-codes/rust-template/issues/1, including:

- [X] CI
  - [X] Runs tests and checks
    - [X] Clippy checks
    - [X] Dependabot file
    - [X] Deployment
- [X] Pre-commit hooks (_Should not run tests_)
  - [X] Clippy
  - [X] Check (`cargo check`), sort dependencies
  - [X] Import ordering and granularity (uses nightly `fmt`)
- [X] README
  - [X] Badges
    - [X] Codecov
    - [X] License
    - [X] Discord (though optional)
- [X] LICENSE
- [X] Codeowners file
- [X] Releases
    - [X] `release-please`
- [X] Nix support
- [X] `wasmpack`

### CODE REVIEW

File: rust+wasm/{{project-name}}-wasm/README.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+- Apache License, Version 2.0, ([LICENSE-APACHE](./LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
+- MIT license ([LICENSE-MIT](./LICENSE-MIT) or http://opensource.org/licenses/MIT)
+
+at your option.
+
+### Contribution
+
+Unless you explicitly state otherwise, any contribution intentionally
+submitted for inclusion in the work by you, as defined in the Apache-2.0
```
Comment (appcypher) [inline rust+wasm/{{project-name}}-wasm/README.md:143]: Missed the conditional here

Comment (zeeshanlakhani) [reply rust+wasm/{{project-name}}-wasm/README.md:143]: This is appropo for the dual license it seems: https://github.com/rustwasm/wasm-pack-template#license.

File: README.md [hunk 2]
```diff
@@ -1 +1,221 @@
...
+      <img src="https://img.shields.io/static/v1?label=Discord&message=join%20us!&color=mediumslateblue" alt="Discord">
+    </a>
+  </p>
+</div>
+
+<div align="center"><sub>:warning: Work in progress :warning:</sub></div>
+
+##
+
+This is Fission's opinionated [Rust][rust] and [Rust][rust]+[WebAssembly][wasm]
```
Comment (bgins) [inline README.md:28]: > This __ is Fission's opinionated

This sentence could use a noun where at the `__`. Maybe "This template" or "This repository"?

File: README.md [hunk 3]
```diff
@@ -1 +1,221 @@
...
+(wasm) project generator, which leverages the [cargo-generate][cargo-generate]
```
Comment (bgins) [inline README.md:29]: > (wasm)

WebAssembly is abbreviated as "Wasm"

File: README.md [hunk 4]
```diff
@@ -1 +1,221 @@
...
+tool.
+
+These templates provide various features for getting-up and running with Rust or
+Rust *and* Wasm, including:
+
+* README standardization, code of conduct, contribuing guidelines, and
+  a consistent project layout
+* GitHub issue and pull-request templates
+* Default Rust dependencies (particulary for wasm)
```
Comment (bgins) [inline README.md:38]:
```suggestion
* Default Rust dependencies (particulary for Wasm)
```

File: README.md [hunk 5]
```diff
@@ -1 +1,221 @@
...
+* Release GitHub Action workflow(s) using the
+  [release-please-action][release-please-action] and the
+  [release-please][release-please] deploy strategy (*optional*)
+  * For wasm libraries, this includes publishing to NPM via
```
Comment (bgins) [inline README.md:42]:
```suggestion
  * For Wasm libraries, this includes publishing to NPM via
```

File: README.md [hunk 6]
```diff
@@ -1 +1,221 @@
...
+    [wasm-pack][wasm-pack], reliant on the Cargo version for the wasm package.
```
Comment (bgins) [inline README.md:43]:
```suggestion
    [wasm-pack][wasm-pack], reliant on the Cargo version for the Wasm package.
```

File: README.md [hunk 7]
```diff
@@ -1 +1,221 @@
...
+  event-based diagnostic information, going beyond just logging-style
+  diagnostics
+- [`tracing-subscriber`][tracing-subscriber] for Rust binary applications
+  to collect trace data, such as by logging it to standard output, and
+  consume messages emitted by log-instrumented libraries and modules.
+
+### Generating a Rust+Wasm Workspace Project
+
+The `rust+wasm` template is designed for generating a workspace containing both
+a rust native library, as well as one compiled for wasm and leveraging
```
Comment (bgins) [inline README.md:117]:
```suggestion
a rust native library, as well as one compiled to Wasm and leveraging
```

This whole "Wasm" not "wasm" thing is a bit nitpicky, so feel free to ignore. The use of "wasm" has been fairly consistent here, and that matters more.

File: README.md [hunk 8]
```diff
@@ -1 +1,221 @@
...
+
+```console
+cargo generate --git https://github.com/fission/rust-template
+```
+
+*Note*: Currently, `wasm-pack` [does not support building binary
+ crates][no-binary], so even with the `--bin` flag specified, a library
+will still be generated.
+
+#### 🔋 Batteries Included
```
Comment (bgins) [inline README.md:130]: Fantastic! Great to mention what is included and why. 💯

File: README.md [hunk 9]
```diff
@@ -1 +1,221 @@
...
+
+## Contributing
+
+### Pre-commit Hook
+
+This repository recommends using [pre-commit][pre-commit] for running pre-commit
+hooks. Please run this before every commit and/or push.
+
+- Once installed, Run `pre-commit install` to setup the pre-commit hooks
+  locally.  This will reduce failed CI builds.
```
Comment (bgins) [inline README.md:164]:
```suggestion
  locally. This will reduce failed CI builds.
```

File: README.md [hunk 10]
```diff
@@ -1 +1,221 @@
...
+
+This repository contains two sub-templates:
+
+* `rust`: for generating a rust-only library or binary/executable project
+* `rust+wasm` for generating a [cargo workspace][cargo-workspace]
+  with a rust-only crate of the project (library or binary) and another crate
+  for wasm-bindings (library-only), meant for execution in [Node.js][node-js]
+  or running in modern browsers and/or with bundlers like [webpack][webpack].
+
+## Getting Started
```
Comment (bgins) [inline README.md:69]: Maybe add a step with instructions for installing `cargo generate` with `cargo install cargo-generate`

**Comment by zeeshanlakhani** [comment]
> In the README's for both templates both `## Getting Help` and `## External Resources` sections are left blank. Just wanted to double check this was intentional

@BoisterousCoder it was since it will be different for those using it. I can put some filter text in though.

File: rust+wasm/{{project-name}}-wasm/LICENSE-APACHE [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+      To apply the Apache License to your work, attach the following
+      boilerplate notice, with the fields enclosed by brackets "[]"
+      replaced with your own identifying information. (Don't include
+      the brackets!)  The text should be enclosed in the appropriate
+      comment syntax for the file format. We also recommend that a
+      file or class name and description of purpose be included on the
+      same "printed page" as the copyright notice for easier
+      identification within third-party archives.
+
+   Copyright [yyyy] [name of copyright owner]
```
Comment (bgins) [inline rust+wasm/{{project-name}}-wasm/LICENSE-APACHE:189]: Would it be worth it to template in the year and copyright owner? Or maybe add them as post-generation tasks?

**Review by bgins — COMMENTED** [review]
Added a few comments. Still testing things out. 👀

File: rust+wasm/{{project-name}}-wasm/LICENSE-APACHE (hunk 11 above)
Comment (zeeshanlakhani) [reply rust+wasm/{{project-name}}-wasm/LICENSE-APACHE:189]: yeah. maybe the org/user can be the owner and copyright we can use by setting the var ourselves. I'll see what I can do. I realize https://github.com/wnfs-wg/rs-wnfs/blob/main/LICENSE#L189 didn't update this haha.

**Review by matheus23 — APPROVED** [review]
After upgrading to latest cargo-generate this is working nicely for me :tada:

**Pull request merged**